The Great Decoupling

pivotal-jwinters opened the pull request:
This PR is based off #3225 (which fixes #2826)

It also fixes #3240 

...and fixes #3267 

...and fixes #2895

vito commented on atc/api/buildserver/abort.go:
```diff
@@ -14,14 +13,7 @@ func (s *Server) AbortBuild(build db.Build) http.Handler {
-		}
-
-		err = engineBuild.Abort(aLog)
+		err := build.MarkAsAborted()
```
Related: https://github.com/concourse/concourse/issues/1585

Should I prioritize that soon while this is fresh in our heads?

pivotal-jwinters replied:
Maybe this could be something to look into when we do the broader LISTEN/NOTIFY work?

vito commented on atc/api/buildserver/server.go:
```diff
@@ -16,9 +15,7 @@ type Server struct {
 	externalURL string
-	peerURL     string
 
-	engine              engine.Engine
```
dope

pivotal-jwinters replied:
yep

vito commented on atc/api/jobserver/create_build.go:
```diff
@@ -33,29 +32,10 @@ func (s *Server) CreateJobBuild(pipeline db.Pipeline) http.Handler {
-		}
-
-		build, _, err := scheduler.TriggerImmediately(logger, job, resources, versionedResourceTypes)
+		build, err := job.CreateBuild()
```
Shouldn't we make sure that this created build is marked 'triggered' so that it gets up-to-date resource versions? As discussed in https://github.com/concourse/concourse/issues/2438

It'd be nice if we kicked the scheduler loop here too. Fast is a feature.

pivotal-jwinters replied:
This already happens here: https://github.com/concourse/concourse/blob/ca9fd7c33fe63291ef291d40c4bd9ccc2ba692d4/atc/db/pipeline.go#L254-L255

vito commented on atc/api/jobserver/list_inputs.go:
```diff
@@ -28,21 +27,13 @@ func (s *Server) ListJobInputs(pipeline db.Pipeline) http.Handler {
 			return
 		}
 
-		err = scheduler.SaveNextInputMapping(logger, job, resources)
```
:+1:

pivotal-jwinters replied:
yep

vito commented on atc/api/pipelineserver/build.go:
```diff
@@ -22,22 +21,13 @@ func (s *Server) CreateBuild(pipelineDB db.Pipeline) http.Handler {
 		}
 
-		build, err := pipelineDB.CreateOneOffBuild()
+		build, err := pipeline.CreateStartedBuild(plan)
```
kicking the build tracker here would be nice

pivotal-jwinters replied:
LISTEN/NOTIFY work?

vito commented on atc/atccmd/command.go:
```diff
@@ -1029,7 +1005,7 @@ func (cmd *RunCommand) defaultBindIP() net.IP {
 }
 
-func (cmd *RunCommand) defaultURL() flag.URL {
+func (cmd *RunCommand) DefaultURL() flag.URL {
```
Should we call this `LocalURL` or something? I always get confused by the name.

pivotal-jwinters replied:
¯\_(ツ)_/¯

vito commented on atc/engine/db_engine.go:
```diff
@@ -15,31 +14,29 @@ import (
 
 const trackLockDuration = time.Minute
 
-func NewDBEngine(engines Engines, peerURL string) Engine {
```
dope

pivotal-jwinters replied:
yep

vito commented on atc/db/migration/migrations/1551384520_fix_build_private_plan.up.go:
```diff
@@ -0,0 +1,76 @@
+	for rows.Next() {
+
+		build := build{}
+		if err = rows.Scan(&build.id, &build.plan, &build.nonce); err != nil {
```
(Same goes for the 'down' migration.)

pivotal-jwinters replied:
heh, yeah that makes sense. I'll batch it.

pivotal-jwinters replied:
I actually remember this now. I don't think this needs to be batched. The `private_plan` only exists for the duration that the build is running. We query for builds where `private_plan` IS NOT NULL so that shouldn't actually be very many things.

pivotal-jwinters replied:
Currently there are only 42 builds on wings where this is true.

vito commented on atc/db/pipeline.go:
```diff
@@ -50,7 +50,10 @@ type Pipeline interface {
 	Builds(page Page) ([]Build, Pagination, error)
+
 	CreateOneOffBuild() (Build, error)
+	CreateStartedBuild(plan atc.Plan) (Build, error)
```
Do we still need `CreateOneOffBuild`?

pivotal-jwinters replied:
ugh, yes. its used absolutely everywhere in the tests and I did not want to mess with that.

vito commented on atc/db/team.go:
```diff
@@ -43,16 +44,20 @@ type Team interface {
 	OrderPipelines([]string) error
 
 	CreateOneOffBuild() (Build, error)
+	CreateStartedBuild(plan atc.Plan) (Build, error)
```
Do we still need `CreateOneOffBuild`?

pivotal-jwinters replied:
so. many. tests.

vito commented on atc/db/pipeline.go:
```diff
@@ -972,6 +975,56 @@ func (p *pipeline) CreateOneOffBuild() (Build, error) {
+		return nil, err
+	}
+
+	err = build.saveEvent(tx, event.Status{
```
Saving an event here feels a little awkward. When does this normally get sent? Will it be sent twice now?

I wonder if it would be better to create it 'pending' and have the tracker also process pending one-off builds?

pivotal-jwinters replied:
I agree, but i think this is a larger discussion around build state. Seems like we need an intermediate `planned` state.

This means the `BuildStarter` actually becomes a `BuildPlanner`.

vito commented on atc/db/team.go:
```diff
@@ -647,6 +677,55 @@ func (t *team) CreateOneOffBuild() (Build, error) {
+		return nil, err
+	}
+
+	err = build.saveEvent(tx, event.Status{
```
same concern as above - i'm a little iffy on inserting these in 'started' state

pivotal-jwinters replied:
RIP BuildStarter, long live BuildPlanner

vito commented on atc/api/artifactserver/create.go:
```diff
@@ -0,0 +1,68 @@
+		// I think leaving the race condition is fine for now. Worst case
+		// is a fly execute will fail and the user will need to rerun it.
+
+		artifact, err := volume.InitializeArtifact("", 0)
```
'create' -> 'initialize' -> 'stream in data' feels unintuitive - normally it's 'initialized' only when it has all its data. it seems like this is to paper over the GC race condition. also why are both arguments zero-values?

pivotal-jwinters replied:
yes yes and yes. we can have this flow make more sense if we figure out the race condition.

lets just pretend those zero values aren't there... they get used for artifacts from build outputs.

vito commented on atc/db/worker_artifact.go:
```diff
@@ -0,0 +1,119 @@
+func (a *artifact) BuildID() int         { return a.buildID }
+func (a *artifact) CreatedAt() time.Time { return a.createdAt }
+
+func (a *artifact) Volume(teamID int) (CreatedVolume, bool, error) {
```
surprised this is given a team ID - will it ever have multiple volumes for different teams? or is this just for checking access?

pivotal-jwinters replied:
yeah its just for checking access. its really unfortunate that teamID exists on the volume itself. It would be a lot nicer and more intuitive if each "higher level resource" had the teamID.

vito commented on atc/exec/task_step.go:
```diff
@@ -440,8 +437,13 @@ func (action *TaskStep) registerOutputs(logger lager.Logger, repository *worker.
+				source := NewTaskArtifactSource(mount.Volume)
+				repository.RegisterSource(artifact.Name(outputName), source)
+
+				_, err := mount.Volume.InitializeArtifact(outputName, action.buildID)
```
So rather than having a step that exists solely to proxy an artifact to the user, we initialize all task outputs as artifacts and just allow them to be downloaded via the API?

Does this mean that each build's task outputs now live for 12 hours? It seems like that could pile up and result in a lot more disk usage if so, assuming a build runs frequently enough.

pivotal-jwinters replied:
TBH I never even considered putting this in its own step (which is hilarious because I had to delete the old output artifact build step to make this change).

It wouldn't really help us with the lifecycle, but it would at least limit the number of `artifacts` that get created to only one-off builds.

vito replied:
Hmm should we bring the artifact output step back, then? I wouldn't want to ship with all task outputs lasting for 12 hours.

pivotal-jwinters replied:
Yeah working on that now. It adds a lot better symmetry :)

vito commented on atc/plan.go:
```diff
@@ -28,7 +28,8 @@ type Plan struct {
+	Name       string `json:"name"`
 }
 
 type ArtifactOutputPlan struct {
```
Should this be removed?

pivotal-jwinters replied:
nah, we still need it

pivotal-jwinters replied:
err I originally thought you we're talking about the `UserArtifactPlan` because of how its highlighted. I now realize you actually meant the ArtifactOutputPlan, which coincidently we are bringing back because of other things you mentioned :)

pivotal-jwinters replied:
glad it never got removed in the first place ;)

vito replied:
lol, yep 👍

vito commented on atc/scheduler/runner.go:
```diff
@@ -22,15 +22,6 @@ type BuildScheduler interface {
 		resourceTypes atc.VersionedResourceTypes,
 	) (map[string]time.Duration, error)
-
-	TriggerImmediately(
```
dope

pivotal-jwinters replied:
yep

vito commented on atc/worker_artifact.go:
```diff
@@ -0,0 +1,8 @@
+	ID        int    `json:"id"`
+	Name      string `json:"name"`
+	BuildID   int    `json:"build_id"`
+	CreatedAt int64  `json:"created_at"`
```
Not actually sure what format we typically use for timestamps in the API, but we should check this to make sure we're consistent.

pivotal-jwinters replied:
from what i saw we aren't consistent lol

vito commented on fly/commands/internal/executehelpers/inputs.go:
```diff
@@ -127,18 +131,30 @@ func TaskInputsContainsName(inputs []atc.TaskInputConfig, name string) bool {
 	kvMap := map[string]Input{}
 
 	for _, i := range inputMappings {
+		artifact, err := Upload(team, i.Path, includeIgnored)
```
I noticed we don't show upload/download progress anymore, now that the steps don't print it. It'd be nice to have that back.

pivotal-jwinters replied:
erm yeah, i'll look into that

pivotal-jwinters replied:
I spoke with @xtremerui and he seems to think this got removed a while ago. If we want that back can we create a separate issue?

vito replied:
It's there for me:

```
~/w/concourse (fix-dashboard-search) $ fly -t ci e -c ci/tasks/docker-compose-watsjs.yml -i concourse=. -p
executing build 4989469 at https://ci.concourse-ci.org/builds/4989469 
initializing
concourse: 16.54 MiB/s 0s                                                       
running concourse/ci/tasks/scripts/with-docker-compose concourse/ci/tasks/scripts/watsjs
/tmp/build/e55deab7/concourse /tmp/build/e55deab7
Creating network "concourse_default" with the default driver
Pulling db (postgres:)...
```

It's not a progress bar, but it still shows the upload rate and duration.

It shouldn't take long to restore it, and I'd like to avoid user-facing regressions on `master` that prevent us from shipping, so I don't want to merge without it. 😅 

You could probably just hoist the code from the user artifact step into `fly`:

https://github.com/concourse/concourse/blob/b20ee6770be162614a2d7e561c5bff095b81db93/atc/exec/user_artifact_step.go#L54-L62

pivotal-jwinters replied:
ah cool! im going to pull this into fly execute instead of having it in the step.

pivotal-jwinters commented on atc/api/buildserver/create.go:
```diff
@@ -22,22 +22,13 @@ func (s *Server) CreateBuild(team db.Team) http.Handler {
-			return
-		}
-
-		go engineBuild.Resume(hLog)
```
yeah can we add this to a general LISTEN/NOTIFY story? I think we could take this approach in a few more places.

vito replied:
created https://github.com/concourse/concourse/issues/3508

pivotal-jwinters commented on atc/api/accessor/accessor_test.go:
```diff
@@ -651,12 +650,16 @@ var _ = Describe("Accessor", func() {
+
+		Entry("owner :: "+atc.GetArtifact, atc.GetArtifact, "owner", true),
+		Entry("member :: "+atc.GetArtifact, atc.GetArtifact, "member", true),
+		Entry("viewer :: "+atc.GetArtifact, atc.GetArtifact, "viewer", true),
```
hmm yeah maybe we should make it member only, there could be sensitive info in these artifacts.

pivotal-jwinters commented on atc/api/artifactserver/create.go:
```diff
@@ -0,0 +1,68 @@
+		// with the way we create volumes for a container
+
+		// I think leaving the race condition is fine for now. Worst case
+		// is a fly execute will fail and the user will need to rerun it.
```
It gets GC'ed because it returns a created volume, with nothing keeping it around. I'll investigate to see if I can do some refactoring to get it to return a creating volume.

vito replied:
Sounds fine.

pivotal-jwinters commented on atc/api/artifactserver/create.go:
```diff
@@ -0,0 +1,68 @@
+			return
+		}
+
+		err = volume.StreamIn("/", r.Body)
```
yep

pivotal-jwinters commented on atc/api/artifactserver/create.go:
```diff
@@ -0,0 +1,68 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+
+		// TODO: can probably check if fly sent us an etag header
```
yep

pivotal-jwinters commented on atc/api/artifactserver/get.go:
```diff
@@ -0,0 +1,66 @@
+		_, err = io.Copy(w, reader)
+		if err != nil {
+			logger.Error("failed-to-encode-artifact", err)
+			w.WriteHeader(http.StatusInternalServerError)
```
not sure what you mean? why does this write multiple headers?

pivotal-jwinters commented:
- [ ] batch migration
- [ ] get artifact -> member only
- [ ] creating volume
- [ ] upload/download status
- [ ] artifact lifecycle

pivotal-jwinters commented:
@vito based on these commits I just pushed, and some comments, I think I addressed all your concerns. Let me know if you think there's something that I missed.

pivotal-jwinters commented:
Also fixed the double status. NOW it should be good to go.
